pkg/rbd/rbd_util.go

```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+		key, rbdSnap.Pool, rbdSnapOMapPrefix+snapUUID, rbdSnapSourceImageKey)
+	if err != nil {
+		if _, ok := err.(util.ErrKeyNotFound); ok {
+			err = unreserveSnap(rbdSnap, credentials)
+		}
+		return false, err
+	}
+
+	// check if snapshot source image omap points back to the source volume passed in
+	if savedVolName != rbdSnap.RbdImageName {
+		// NOTE: This can happen if there is a snapname conflict, and we alerady have a snapshot
+		// with the same name pointing to a different RBD image as the source
+		err = fmt.Errorf("snapname points to different image, request name (%s)"+
+			" image name (%s) image omap"+" volume name (%s)",
+			rbdSnap.RequestName, rbdSnap.RbdImageName, savedVolName)
+		return false, ErrSnapNameConflict{rbdSnap.RequestName, err}
+	}
+
+	// Fetch on-disk image attributes
+	err = updateSnapWithImageInfo(rbdSnap, credentials)
+	if err != nil {
+		if _, ok := err.(ErrSnapNotFound); ok {
+			err = unreserveSnap(rbdSnap, credentials)
+			return false, err
+		}
+
+		return false, err
+	}
+
+	// found a snapshot already available, process and return its information
+	poolID, err := util.GetPoolID(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool)
+	if err != nil {
+		klog.Errorf("internal error fetching pool ID (%s)", err)
+		return false, err
+	}
+
+	vi := util.CsiIdentifier{
+		PoolID:          poolID,
+		EncodingVersion: volIDVersion,
+		ClusterID:       rbdSnap.ClusterID,
+		ObjectUUID:      snapUUID,
+	}
+	rbdSnap.SnapID, err = vi.ComposeCsiID()
+	if err != nil {
+		return false, err
+	}
+
+	klog.V(4).Infof("Found existing snap (%s) with snap name (%s) for request (%s)",
+		rbdSnap.SnapID, rbdSnap.RbdSnapName, rbdSnap.RequestName)
+
+	return true, nil
+}
+
+// ErrVolNameConflict is generated when a requested CSI volume name already exists on RBD but with
+// different properties, and hence is in conflict with the passed in CSI volume name
+type ErrVolNameConflict struct {
+	requestName string
+	err         error
+}
+
+func (e ErrVolNameConflict) Error() string {
+	return e.err.Error()
+}
+
+/*
+Check comment on checkSnapExists, to understand how this function behaves
+
+**NOTE:** These functions manipulate the rados omaps that hold information regarding
+volume names as requested by the CSI drivers. Hence, these need to be invoked only when the
+respective CSI snapshot or volume name based locks are held, as otherwise racy access to these
+omaps may end up leaving the omaps in an inconsistent state.
+*/
+func checkVolExists(rbdVol *rbdVolume, credentials map[string]string) (found bool, err error) {
+	var vi util.CsiIdentifier
+
+	// Structure members used to determine if provided rbdVolume exists, are checked here, to
+	// provide an easy way to check when this function can be reused
+	if rbdVol.RequestName == "" || rbdVol.Monitors == "" || rbdVol.AdminID == "" ||
+		rbdVol.Pool == "" || rbdVol.ClusterID == "" || rbdVol.VolSize == 0 {
+		return false, errors.New("missing information in rbdVolume to check for existence")
```

Comment:
This if condition looks bit ugly to me.

Reply:
Checking each element and printing what is missing can be tedious, this is a developer log when the function is reused. Hence, printed the entire structure contents on errors, to help aid debugging instead.

Reply:
Addressed now with a validate routine for rbdSnap and rbdVol to throw exact field that is missing and handle the error.